src/EnergyPlus/SetPointManager.hh

```diff
@@ -955,240 +505,77 @@ namespace SetPointManager {
-    Array1D<SetPointManager::DefRABFlowSetPointManager> RABFlowSetPtMgr;                          // Array for return air bypass
-    Array1D<SetPointManager::DefMultiZoneAverageCoolingSetPointManager> MZAverageCoolingSetPtMgr; // Array for MultiZone
-    Array1D<SetPointManager::DefMultiZoneAverageHeatingSetPointManager> MZAverageHeatingSetPtMgr; // Array for MultiZone
-    Array1D<SetPointManager::DefMultiZoneAverageMinHumSetPointManager> MZAverageMinHumSetPtMgr;   // Array for MultiZone
-    Array1D<SetPointManager::DefMultiZoneAverageMaxHumSetPointManager> MZAverageMaxHumSetPtMgr;   // Array for MultiZone
-    Array1D<SetPointManager::DefMultiZoneMinHumSetPointManager> MZMinHumSetPtMgr;                 // Multizone min humidity rat Set Pt Mgr
-    Array1D<SetPointManager::DefMultiZoneMaxHumSetPointManager> MZMaxHumSetPtMgr;                 // Multizone max humidity rat Set Pt Mgr
-    Array1D<SetPointManager::DefineFollowOATempSetPointManager> FollowOATempSetPtMgr;             // Array for Follow Outdoor Air
-    Array1D<SetPointManager::DefineFollowSysNodeTempSetPointManager> FollowSysNodeTempSetPtMgr;   // Array for Follow System
-    Array1D<SetPointManager::DefineGroundTempSetPointManager> GroundTempSetPtMgr;                 // Array for Ground Temp Setpoint
-    Array1D<SetPointManager::DefineCondEntSetPointManager> CondEntSetPtMgr;                       // Condenser Entering Water Set Pt Mgr
-    Array1D<SetPointManager::DefineIdealCondEntSetPointManager> IdealCondEntSetPtMgr;             // Ideal Condenser Entering Set Pt Mgr
-    Array1D<SetPointManager::DefineSZOneStageCoolinggSetPointManager> SZOneStageCoolingSetPtMgr;  // single zone 1 stage cool
-    Array1D<SetPointManager::DefineSZOneStageHeatingSetPointManager> SZOneStageHeatingSetPtMgr;   // single zone 1 stage heat
-    Array1D<SetPointManager::DefineReturnWaterChWSetPointManager> ReturnWaterResetChWSetPtMgr;    // return water reset
-    Array1D<SetPointManager::DefineReturnWaterHWSetPointManager> ReturnWaterResetHWSetPtMgr;      // hot-water return water reset
-    Array1D<SetPointManager::DefineScheduledTESSetPointManager> SchTESSetPtMgr;                   // Array for TES Scheduled Setpoint Manager data
-    Array1D<SetPointManager::DefineSysNodeResetSetPointManager> SystemNodeResetSetPtMgr;          // Array for System Node Reset Setpoint Manager data
-
-    Real64 CondWaterSetPoint = 0; // Condenser entering water temperature setpoint this timestep, C
-    Real64 EvapOutletTemp = 0;    // Evaporator water outlet temperature (C)
-    Real64 CondTempLimit = 0;     // Condenser entering water temperature setpoint lower limit
-    Real64 CurLoad = 0;           // Current cooling load, W
-    Real64 TotEnergy = 0;         // Total energy consumptions at this time step
-    Real64 TotEnergyPre = 0;      // Total energy consumptions at the previous time step
-
+    Array1D<SetPointManager::SPMBase *> spms;
```

Comment:
"But performance" you may say. well,

https://quick-bench.com/q/9xIjZoQqSyiz-uhlVFTFpN7fsn0

Reply:
> I'm pretty sure you picked a raw pointer on purpose after consideration, but in the off chance not, I'd prefer using `std::unique_ptr<SPMBase>`. (And maybe a std::array or EPVector to contain it?)

I did use a raw pointer on consideration. What is the consideration for `std::unique_pointer`?EnergyPlus is not a traditional scientific program, but it does have the same simple memory management  pattern as scientific programs do. You allocate everything at the beginning, then you compute, then you deallocate. There's no need for reference counting and garbage collection in that model. Just add `delete` loops to `clear_state`. To me, `std::unique_pointer` is just one more box in C++-library bingo.

Also, I'm out on `EPVector`. I no longer see the point and in fact I am convinced that a container that can be accessed as both a 0-based and 1-based array is only a source of bugs. I could change to `std::vector` and I have done that in some modules that are more self-contained than `SPM` (e.g., `OutputProcessor`). It will probably take another day or so to dig up the corner cases that arise. Is that worth it for this PR?

Reply:
> "But performance" you may say. well,
> 
> https://quick-bench.com/q/9xIjZoQqSyiz-uhlVFTFpN7fsn0
> 
> Locally I get more similar results, but still, my point is that this does not hinder performance.
> 
> ![image](https://private-user-images.githubusercontent.com/5479063/336021802-3a61ba2a-0cc9-4304-ba31-a7d9f1b659f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc0MTk1MjQsIm5iZiI6MTcxNzQxOTIyNCwicGF0aCI6Ii81NDc5MDYzLzMzNjAyMTgwMi0zYTYxYmEyYS0wY2M5LTQzMDQtYmEzMS1hN2Q5ZjFiNjU5ZjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDNUMTI1MzQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzEwMzhkMTdlMGRlYTI5NjkwNDM3YmM1YTI2NzNkY2ZlYWFiY2Y2MWVkYTAzNmJhN2VjNWVlMDFhZWRhZjlhMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Zd8UHWPfAHJRCXWpyd5CyZiqdzisd9rSoboFIuxQkaQ)

I will not say "but performance". I will say "C++ bingo".

Reply:
Regarding `EPVector`, the original intent was as a shim to move toward `std::vector` and `std::array` and not as something we should keep around forever. With `EPVector`, we wanted something that would work well enough with `Array1D` (which allows 0-based and 1-based indexing and is the reason that `EPVector` does it) that switching wouldn't break everything.

New code should definitely use `std::vector/array` unless there's a compelling reason to go another direction. At the time that `EPVector` was introduced, there were a lot of problematic cases (slicing etc.) that are less present now. Anything that's an `EPVector` now should (hopefully) just need the indexing addressed in order to be switched.

Reply:
I agree. I think this is why I changed my mind about it. I was pro `EPVector` earlier because there was just a lot of `Array1D` madness that we needed to move away from. But now that we've moved away from a lot of it, the 0-based/1-based duality has become somewhat of a bug rather than a feature for external references (I was bitten by it a few times during the `Fans` transition). FWIW I actually think that Array1D also allows mixed indexing to some degree and we should disable it. 

I also agree that we need to move to 0-based indices incrementally and I was thinking of doing that here, but it took me a little while to actually work the kinks out of this module and by the time I was done I was too tired to deal with that type of change. I will probably go back and do it in a separate pass.

Reply:
There used to be way too much slicing going on, that's for sure. Let's not add any new `Array1D` objects if we can help it. With Fortran indexing being what it is (negative indices and all that), disabling either `operator()` or `operator[]` on `Array1D` might be harder than just moving to `EPVector` and then on to `std::vector`.

Reply:
I was thinking of disabling `operator []` on it. I don't think it's used in more than a handful of places, maybe I'm wrong.